### ISSUE TITLE
Context for global functions

### ISSUE THREAD

**Issue by max-frai** [Keats/tera#299]
Hello, I use fluent-rs for internationalization.
For now tera supports to register only global static functions without context which help process incoming data, but what if I need some additional context data?
In template:
`{{ global_lang(key_name) }}`
In code:
`context.add("key_name", "localization-key-name")`

So I want to call `global_lang` with `localization-key-name` string as argument and inside access external object which stores all locale keys, etc and return result string data.

**Comment by max-frai** [comment]
Seems like this is related #219

**Comment by Keats** [comment]
I'd be happy to have a PR for #219, especially if it can fix your issues

**Comment by andy128k** [comment]
`Arc` to the rescue. [This example](https://github.com/Keats/tera/blob/v1/src/renderer/tests/basic.rs#L672) shows how to deal with it. What do I miss here?

**Comment by Keats** [comment]
`Next` doesn't contain a pointer. Here is an example of a Zola fn:

```rust
#[derive(Debug)]
pub struct GetPage<'a> {
    pages: Arc<HashMap<&'a String, SerializingPage<'a>>>,
}
impl<'a> GetPage<'a> {
    pub fn new(library: &'a Library) -> Self {
        let mut pages = HashMap::new();
        for page in library.pages_values() {
            pages.insert(
                &page.file.relative,
                library.get_page(&page.file.path).unwrap().to_serialized(library),
            );
        }

        Self {pages: Arc::new(pages)}
    }
}
impl<'a> Function for GetPage<'a> {
    fn call(&self, args: &HashMap<String, Value>) -> Result<Value> {
        let path = required_arg!(
            String,
            args.get("path"),
            "`get_page` requires a `path` argument with a string value"
        );
        match self.pages.get(&path) {
            Some(p) => Ok(to_value(p).unwrap()),
            None => Err(format!("Page `{}` not found.", path).into()),
        }
    }
}
```
Trying to register it gives a conflicting lifetime error because:

```rust
    pub fn register_tera_global_fns(&mut self) {
        self.tera.register_function("trans", global_fns::Trans::new(self.config.clone()));
        self.tera.register_function("get_page", global_fns::GetPage::new(&self.library));
}
```
The lifetime of `&self.library` will end and won't outlive the `'static`. I don't know if there is a way to fix that tbh

**Comment by Keats** [comment]
If we can tie the lifetime to the traits to the Tera instance instead of `'static`, it might work?

**Comment by andy128k** [comment]
Giving a lifetime to an instance of `Tera` could help. Have a look at my attempt here https://github.com/Keats/tera/pull/378 . I don't know how it fits into Zola's code, but I'm afraid with references ergonomics could be worse.

**Comment by technic** [comment]
Hi! Why do we need to use an external variable and the synchronization for this? Can't we pass reference to `Context` inside each function?

Synchronization will mean, that we can not render templates on web server in parallel. Because each client needs different value of the global variable. (for example different languages).

**Comment by Keats** [comment]
Are you referring to the content of that issue or the current version?

**Comment by technic** [comment]
Sorry, I didn't understand your question. I mean `tera::Context`.

Basically I want something like this:

```html
<p> translate("foobar") <p>
```

```rust
let mut ctx = Context::new();
ctx.insert("LANG", "en");
t.render(template_name, ctx)
```

so basically lambdas that we use in `register_function` in addition to `HashMap` with arguments also accepts `&Context` reference.
Then I can now value of "LANG" inside `translate` implementation.

**Comment by technic** [comment]
@Keats I created a new issue with more explanation https://github.com/Keats/tera/issues/543.
